engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java

```diff
@@ -349,22 +340,11 @@ else if ( executableReturnValue != null ) {
 	}
 
 	public boolean hasMetaConstraintBeenProcessed(Object bean, Path path, MetaConstraint<?> metaConstraint) {
-		// TODO switch to proper multi key map (HF)
-		IdentitySet processedConstraints = processedMetaConstraints.get( new BeanAndPath( bean, path ) );
-		return processedConstraints != null && processedConstraints.contains( metaConstraint );
+		return processedUnits.contains( new BeanPathMetaConstraintProcessedUnit( bean, path, metaConstraint ) );
```

Comment:
this line bothers me a bit. From chat conversations we know the contains() is problematic from a performance perspective and I'm inclined to suspect the allocation of `BeanPathMetaConstraintProcessedUnit`.

Would be interesting to see if the performance improves more by having nested maps rather than a Set with composite keys.

Reply:
That's basically what was done before (at least, it was done partly). And the creation/allocations/resizing of the nested structures made it worse.

I can try it again nesting everything though.